File: Typecast/TCFpgmTable.m
//
//  TCFpgmTable.m
//  Type Designer
//
//  Created by David Schweinsberg on 30/12/12.
//  Copyright (c) 2012 David Schweinsberg. All rights reserved.
//

#import "TCFpgmTable.h"
#import "Type_Designer-Swift.h"
#import "TCDisassembler.h"

@implementation TCFpgmTable

- (id)initWithDataInput:(TCDataInput *)dataInput directoryEntry:(TCDirectoryEntry *)entry
{
    self = [super init];
    if (self)
    {
        self.directoryEntry = [entry copy];
        [self readInstructionsWithDataInput:dataInput length:[entry length]];
    }
    return self;
}

- (uint32_t)type
{
    return TCTable_fpgm;
}

- (NSString *)name
{
    uint32_t type = [self type];
    return [NSString stringWithFormat:
            @"%c%c%c%c",
            (char)(type >> 24),
            (char)(type >> 16),
            (char)(type >> 8),
            (char)type];
}

- (NSString *)description
{
    return [TCDisassembler disassembleInstructions:[self instructions]
                                 leadingSpaceCount:0];
}

@end

File: Typecast/TCFpgmTable.swift
//
//  TCFpgmTable.swift
//  Type Designer
//
//  Created by David Schweinsberg on 8/3/17.
//  Copyright © 2017 David Schweinsberg. All rights reserved.
//

import Foundation
import IOUtils

class TCFpgmTable: TCBaseTable, Codable {
  var instructions = [UInt8]()

  init(data: Data) {
    let dataInput = TCDataInput(data: data)
    instructions = dataInput.read(length: data.count)
  }

  override class var tag: UInt32 {
    get {
      return TCTableTag.fpgm.rawValue
    }
  }

  override var name: String {
    get {
      let tag = type(of: self).tag
      return String(format: "%c%c%c%c",
                    CChar(truncatingIfNeeded:tag >> 24),
                    CChar(truncatingIfNeeded:tag >> 16),
                    CChar(truncatingIfNeeded:tag >> 8),
                    CChar(truncatingIfNeeded:tag))
    }
  }

  override var description: String {
    get {
      return TCDisassembler.disassemble(instructions: instructions,
                                        leadingSpaceCount: 0)
    }
  }
}
